sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/rel/BeamRelNode.java

```diff
@@ -17,15 +17,77 @@
  */
 package org.apache.beam.sdk.extensions.sql.impl.rel;
 
+import java.util.List;
 import org.apache.beam.sdk.transforms.PTransform;
 import org.apache.beam.sdk.values.PCollection;
 import org.apache.beam.sdk.values.PCollectionTuple;
 import org.apache.beam.sdk.values.Row;
 import org.apache.calcite.rel.RelNode;
+import org.apache.calcite.rel.type.RelDataType;
+import org.apache.calcite.rel.type.RelDataTypeField;
 
 /** A {@link RelNode} that can also give a {@link PTransform} that implements the expression. */
 public interface BeamRelNode extends RelNode {
 
+  default double estimateRowSize(RelDataType rowType) {
+    List<RelDataTypeField> fieldList = rowType.getFieldList();
+
+    return fieldList
+        .stream()
+        .mapToDouble(typeField -> estimateDataTypeSize(typeField.getType()))
+        .sum();
+  }
+
+  default double estimateDataTypeSize(RelDataType type) {
+    switch (type.getSqlTypeName()) {
+      case TINYINT:
+        return 1;
+      case SMALLINT:
+        return 2;
+      case INTEGER:
+        return 4;
+      case BIGINT:
+        return 8;
+      case BOOLEAN:
+        return 1;
+      case FLOAT:
+        return 4;
+      case DOUBLE:
+        return 8;
+      case VARCHAR:
+        return 12;
+      case CHAR:
+        return 1;
```

Comment:
Actually in SQL `CHAR` just means "fixed-length string". We don't track the lengths anyhow. And if you switch this to the `RowCoder` then it will just look at `STRING`.